src/utils/headers.test.js

```diff
@@ -83,4 +110,54 @@ test('_headers: rulesForPath testing', (t) => {
     'X-Frame-Options': ['SAMEORIGIN'],
     'X-Frame-Thing': ['SAMEORIGIN'],
   })
+  t.deepEqual(objectForPath(rules, '/placeholder/_next/static/chunks/placeholder'), {
+    'X-Frame-Thing': ['SAMEORIGIN'],
+    'cache-control': ['public', 'max-age=31536000', 'immutable'],
+  })
+  t.deepEqual(objectForPath(rules, '/directory/placeholder/test.html'), {
+    'X-Frame-Thing': ['SAMEORIGIN'],
+    'X-Frame-Options': ['test'],
+  })
+})
+
+/**
+ * The bulk of the _headers logic concerns testing whether or not a path matches
+ * a rule - focus on testing `matchPaths` over `objectForPath` (the latter just
+ * straightforwardly combines a bunch of objects).
+ */
+test('_headers: matchPaths matches rules as expected', (t) => {
+  /**
+   * Make sure (:placeholder) will NOT match root dir.
+   */
+  t.assert(!matchPaths('/:placeholder', '/'))
+  /**
+   * (:placeholder) wildcard tests.
+   */
+  t.assert(matchPaths('/directory/:placeholder', '/directory/test'))
+  t.assert(matchPaths('/directory/:placeholder/test', '/directory/placeholder/test'))
+  t.assert(!matchPaths('/directory/:placeholder', '/directory/test/test'))
+  /**
+   * (*) wildcard tests.
+   */
+  t.assert(matchPaths('/path/*/dir', '/path/to/dir'))
+  t.assert(matchPaths('/path/to/*/*/*', '/path/to/one/two/three'))
+  t.assert(matchPaths('/path/*/to/*/dir', '/path/placeholder/to/placeholder/dir'))
+  t.assert(!matchPaths('/path/*/to/*/dir', '/path/placeholder/to/placeholder'))
+  /**
+   * Trailing (*) wildcard matches recursive subdirs.
+   */
+  t.assert(matchPaths('/path/*', '/path/placeholder/to/placeholder/dir'))
+  t.assert(matchPaths('/path/to/*', '/path/to/oneDir'))
+  t.assert(matchPaths('/path/to/*', '/path/to/oneDir/twoDir/threeDir'))
+  /**
+   * Trailing (*) wildcard matches parent dir. This is caught automagically by
+   * index >= pathParts.length && index >= ruleParts.length check.
+   */
+  t.assert(matchPaths('/path/to/dir/*', '/path/to/dir'))
+  /**
+   * Mixed (*) and (:placeholder) wildcards.
+   */
+  t.assert(matchPaths('/path/*/to/:placeholder/:placeholder/*', '/path/placeholder/to/placeholder/dir/test'))
+  t.assert(matchPaths('/path/*/:placeholder', '/path/to/dir'))
+  t.assert(matchPaths('/path/:placeholder/:placeholder/*', '/path/to/dir/one/two/three'))
 })
```

Comment:
Headers tests should focus on testing `matchPaths`, since that is where the algorithm logic is. I tried to cover almost all cases here. Please let me know if I missed any.